runtime/src/read_optimized_dashmap.rs

```diff
@@ -0,0 +1,199 @@
+#![allow(dead_code)]
+
+#[cfg(feature = "shuttle-test")]
+use shuttle::sync::Arc;
+#[cfg(not(feature = "shuttle-test"))]
+use std::sync::Arc;
+use {
+    dashmap::{
+        mapref::{entry::Entry, multiple::RefMulti},
+        DashMap,
+    },
+    std::{hash::BuildHasher, ops::Deref},
+};
+
+type DashmapIteratorItem<'a, K, V, S> = RefMulti<'a, K, ROValue<V>, S>;
+
+// Wrapper around dashmap that stores (K, Arc<V>) to minimize shard contention.
+#[derive(Debug)]
+pub struct ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    inner: DashMap<K, ROValue<V>, S>,
+}
+
+impl<K, V, S> ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    pub fn new(inner: DashMap<K, ROValue<V>, S>) -> Self {
+        Self { inner }
+    }
+
+    /// Alternative to entry(k).or_insert_with(default) that returns an Arc<V> instead of returning a
+    /// guard that holds the underlying shard's write lock.
+    ///
+    /// # Safety
+    ///
+    /// - Care must be taken when mutating the returned value. If modifications are applied while
+    ///   another thread concurrently deletes the key, the changes may be lost.
+    pub unsafe fn get_or_insert_with(&self, k: &K, default: impl FnOnce() -> V) -> ROValue<V> {
+        match self.inner.get(k) {
+            Some(v) => ROValue::clone(&*v),
+            None => ROValue::clone(
+                self.inner
+                    .entry(k.clone())
+                    .or_insert_with(|| ROValue::new(default()))
+                    .value(),
+            ),
+        }
```

Comment:
I'm assuming the idea here is to grab the read lock (avoid grabbing write lock) if we have a value here.
Otherwise, grab the write lock, perform another lookup, check for presence (because this can race), insert if necessary, and grab the value.

Does something like this work to avoid the double look-up?
```
let entry = self.inner.entry_ref(&k).or_insert_with(|| ROValue::new(default()));
ROValue::clone(e.value())
```

Reply:
I'm hallucinating again. Somehow wandered into a completely unrelated crate. Apologies